### PR TITLE
Add documentation of StaticFinalFieldFolding

### PR DESCRIPTION
**Pull request by liqunl**
Part of #11974

[ci skip]

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
@0xdaryl May I have your review?

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:39]: field -> fields

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:40]: For a primitive type, ...

For a reference type, ...

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:44]: is trying to fold

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:55]: I think:  "...happens, the JIT code with the invalid folded value must not be executed."

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:57]: guard -> guards

situation -> situations

File: runtime/compiler/optimizer/StaticFinalFieldFolding.hpp
Comment (0xdaryl) [inline runtime/compiler/optimizer/StaticFinalFieldFolding.hpp:61]: place -> placed      ?

**Review by 0xdaryl — COMMENTED** [review]
Thanks for writing this.  Only a few minor typo corrections from me.

**Comment by liqunl** [comment]
@0xdaryl Made some changes according to your comments, could you review again?

**Comment by 0xdaryl** [comment]
Thanks.  I'll wait for @vijaysun-omr before merging.

**Comment by vijaysun-omr** [comment]
Can it be clearly stated whether both 1 and 2 foldings are enabled by default ? If so, at which opt levels or under which conditions ?

**Comment by vijaysun-omr** [comment]
I am fine with the rest of the document here (thanks @liqunl for the well written description) and once my prior comment is addressed please feel free to merge it without waiting for any formal approval from me.

**Comment by liqunl** [comment]
Added the conditions when the two types of folding happen to the end of the documentation. Also made a correction to the statement that the folding util can be called anytime in the optimizer. It can only be called before OSRGuardInsertion.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**
